scipy/signal/_upfirdn_apply.pyx

```diff
@@ -0,0 +1,114 @@
+# -*- coding: utf-8 -*-
+
+# Code adapted from "upfirdn" python library with permission:
+#
+# Copyright (c) 2009, Motorola, Inc
+#
+# All Rights Reserved.
+#
+# Redistribution and use in source and binary forms, with or without
+# modification, are permitted provided that the following conditions are
+# met:
+#
+# * Redistributions of source code must retain the above copyright notice,
+# this list of conditions and the following disclaimer.
+#
+# * Redistributions in binary form must reproduce the above copyright
+# notice, this list of conditions and the following disclaimer in the
+# documentation and/or other materials provided with the distribution.
+#
+# * Neither the name of Motorola nor the names of its contributors may be
+# used to endorse or promote products derived from this software without
+# specific prior written permission.
+#
+# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS
+# IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO,
+# THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+# PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR
+# CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+# EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+# PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+# PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
+# LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
+# NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+# SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+cimport cython
+cimport numpy as np
+import numpy as np
+
+
+ctypedef double complex double_complex
+ctypedef float complex float_complex
+
+ctypedef fused DTYPE_t:
+    # Eventually we could add "object", too, but then we'd lose the "nogil"
+    # on the _apply_impl function.
+    float
+    float_complex
+    double
+    double_complex
+
+
+def _output_len(Py_ssize_t len_h,
+                Py_ssize_t in_len,
+                Py_ssize_t up,
+                Py_ssize_t down):
+    """The output length that results from a given input"""
+    cdef Py_ssize_t np
+    cdef Py_ssize_t in_len_copy
+    in_len_copy = in_len + (len_h + (-len_h % up)) // up - 1
+    np = in_len_copy * up
+    cdef Py_ssize_t need = np // down
+    if np % down > 0:
+        need += 1
+    return need
+
+
+def _apply(DTYPE_t [:] x, DTYPE_t [:] h_trans_flip, DTYPE_t [:] out,
+                 Py_ssize_t up, Py_ssize_t down):
+    _apply_impl(x, h_trans_flip, out, up, down)
+
+
+@cython.cdivision(True)  # faster modulo
```

Comment:
@sturlamolden I left this one in because it should make the code a tiny bit faster, as we are guaranteed the variables under modulus operations are positive. Make sense, or no?